Add 2024 unsafe functions (#304)
rust-lang/edition-guide: src/rust-2024/unsafe-functions.md

# Unsafe functions

🚧 The 2024 Edition has not yet been released and hence this section is still "under construction".
More information may be found in the tracking issue at <https://github.com/rust-lang/rust/issues/124866>.

## Summary

- The following functions are now marked [`unsafe`]:
    - [`std::env::set_var`]
    - [`std::env::remove_var`]

[`unsafe`]: ../../reference/unsafe-keyword.html#unsafe-functions-unsafe-fn
[`std::env::set_var`]: ../../std/env/fn.set_var.html
[`std::env::remove_var`]: ../../std/env/fn.remove_var.html

## Details

It can be unsound to call [`std::env::set_var`] or [`std::env::remove_var`] in a multi-threaded program due to safety limitations of the way the process environment is handled on some platforms. The standard library originally defined these as safe functions, but it was later determined that was not correct.

> It is important to ensure that these functions are not called in a multi-threaded program. See the [Safety] section of the function documentation for more details.

Reviewer: ```suggestion
It is important to ensure that these functions are not called when any other thread might be writing to or reading from the environment. See the [Safety] section of the function documentation for more details.
```

Ordinarily it would be a backwards-incompatible change to add `unsafe` to these functions. To address that problem, they are marked as `unsafe` only in the 2024 Edition.

[Safety]: ../../std/env/fn.set_var.html#safety

## Migration

> To make your code compile in both the 2021 and 2024 editions, you will need to make sure that `set_var` and `remove_var` is called from an `unsafe` blocks.

Reviewer: ```suggestion
To make your code compile in both the 2021 and 2024 editions, you will need to make sure that both `set_var` and `remove_var` are called only from within `unsafe` blocks.
```
